src/modm/driver/usb/stusb4500.hpp

```diff
@@ -0,0 +1,157 @@
+		// CcStatus = 0x11,
+		// TxHeaderLow = 0x51,
+		// PdCommandCtrl = 0x1A,
+		DpmPdoNumb = 0x70,
+		DpmSnkPdo1 = 0x85,
+		DpmSnkPdo2 = 0x89,
+		DpmSnkPdo3 = 0x8D,
+	};
+
+	struct RdoRegStatusData {
+		uint32_t MaxCurrent; /// [mA]
+		uint32_t OperatingCurrent; /// [mA]
+		uint8_t ObjectPos;
+	};
+};
+
+/**
+ * @ingroup modm_driver_stusb4500
+ * @author Henrik Hose, Raphael Lehmann
+ */
+template<class I2cMaster>
+class Stusb4500 : public stusb4500, public modm::I2cDevice< I2cMaster, 4 >
+{
+public:
+	Stusb4500(uint8_t address=0x28):
+		modm::I2cDevice<I2cMaster,4>(address)
+	{};
+
+	/**
+	 * Configures a PDO.
+	 *
+	 * Note: According to the USB standard the first PDO (1) always has to set
+	 *  voltage to 5000 [mV].
+	 *
+	 * @param	pdoNumber Which PDO to configure, range 1 to 3.
+	 * @param	voltage	  in mV
+	 * @param	current	  in mA
+	 */
+	modm::ResumableResult<bool>
+	configurePdo(uint8_t pdoNumber, uint32_t voltage, uint32_t current)
+	{
+		RF_BEGIN();
+
+		if((pdoNumber < 1) || (pdoNumber > 3))
+			RF_RETURN(false);
+
+		RF_CALL(readRegister(DpmSnkPdos[pdoNumber - 1], buffer2.data(), 4));
+
+		current /= 10;
+		voltage /= 50;
+
+		buffer2[0] = current;
+		buffer2[1] = ((current >> 8) & 0b11) | ((voltage & 0b11'1111 ) << 2);
+		buffer2[2] = (buffer2[2] & 0b1111'0000) | ((voltage >> 6) & 0b1111);
+
+		RF_END_RETURN_CALL(updateRegister(DpmSnkPdos[pdoNumber - 1], buffer2.data(), 4));
+	}
+
+	/**
+	 * Sets the valid PDO.
+	 *
+	 * @param	pdoNumber Which PDO to set valid, range 1 to 3.
+	 */
+	modm::ResumableResult<bool>
+	setValidPdo(uint8_t pdoNumber)
+	{
+		RF_BEGIN();
+
+		if((pdoNumber < 1) || (pdoNumber > 3))
+			RF_RETURN(false);
+
+		RF_END_RETURN_CALL(updateRegister(Register::DpmPdoNumb, &pdoNumber, 1));
+	}
+
+	/**
+	 * Retrieve RDO status from STUSB4500 chip.
+	 *
+	 * @return	RdoRegStatusData object. Typically only
+	 *  RdoRegStatusData::MaxCurrent [mA] is of interest.
+	 */
+	modm::ResumableResult<RdoRegStatusData>
+	getRdoRegStatus()
+	{
+		RF_BEGIN();
+
+		RF_CALL(readRegister(Register::RdoRegStatus, buffer2.data(), 4));
+
+		// MaxCurrent: Bits 9..0; OperatingCurrent: Bits 19..10; ObjectPos: Bits 30..28
+		rdoRegStatusData.MaxCurrent = buffer2[0] | ((buffer2[1] & 0b11) << 8);
+		rdoRegStatusData.MaxCurrent *= 10;
+		rdoRegStatusData.OperatingCurrent = ((buffer2[1] & 0b1111'1100) >> 2) | ((buffer2[2] & 0b1111) << 6);
+		rdoRegStatusData.OperatingCurrent *= 10;
+		rdoRegStatusData.ObjectPos = (buffer2[3] & 0b0110'0000) >> 5;
+
+		RF_END_RETURN(rdoRegStatusData);
+	}
+
+public:
+	modm::ResumableResult<bool>
+	readRegister(Register reg, uint8_t* output, size_t length)
+	{
+		RF_BEGIN();
+		buffer[0] = uint8_t(reg);
+		this->transaction.configureWriteRead(buffer.data(), 1, output, length);
+		RF_END_RETURN_CALL( this->runTransaction() );
+	}
+
+	modm::ResumableResult<bool>
+	updateRegister(Register reg, const uint8_t* data, size_t length)
+	{
+		RF_BEGIN();
+		if(length > (std::tuple_size<decltype(buffer)>::value - 1))
```

Comment:
You wrong. [Explain this](https://godbolt.org/z/Wzh178) and read the specifications.

Reply:
See https://godbolt.org/z/58MTr7

Reply:
@jasa It does not work here because `buffer` is a non-static class member. The `this` pointer is not a constant expression here. Therefore, calling a method on a non-static member never can be: https://godbolt.org/z/bfnh1f

Reply:
LOL, this is like the Uno reverse card of reviews: No, @jasa, **YOU** WRONG. 😂🤣 C++ is weird…

Reply:
~But here we could use `buffer.size()` since this check now is 100% runtime.~

Using `std::tuple_size<decltype(buffer)>::value` generates faster and smaller code: https://godbolt.org/z/GKMMY8

![grafik](https://user-images.githubusercontent.com/2820734/110630872-38a7b000-81a6-11eb-8f26-a466ed87947a.png)

Reply:
I'm blind, sorry.

Reply:
We don't compile with `-O0` here :wink:

Reply:
Should I change it to `buffer.size()` (and thus trigger the CI again)?

Reply:
I can't tell you about what the CI can take, but `buffer.size()` is by far more readable and idiomatic code.